### PR TITLE
ONNX cherry picks for 1.12.1

### LINKED ISSUE #82275: [ONNX] Cherry pick quantization for 1.12.1 release
`git log --oneline  master ^"justinchu/1.12.1-onnx-quantization" --grep "ONNX"`

```
b7f9315eac [ONNX] Export aten::native_dropout (#81743)
6ea422dd0b Format torch/onnx with ufmt (#82137)
6691ed7884 (upstream/gh/ezyang/1282/base) [ONNX] Export aten::_log_softmax (#81804)
e32691dc7a [ONNX] extend add and sub exporter to cover graph non-tensor inputs (#81736)
1a9317ca64 [ONNX] Export aten::convolution (#81815)
d30784be31 (upstream/gh/ezyang/1275/base) [ONNX] Fix ONNX aten::mul exporter with boolean inputs (#81671)
4c0000a98e [ONNX] Remove duplicated test run (#81146)
df4226b48a [ONNX] Fix bug using std::copy_if (#80999)
74c6a754a5 [ONNX] Deprecation utility (#81901)
4c728a7581 (upstream/gh/SherlockNoMad/3/base) [ONNX] Add tests to quantized cat (#81484)
69608fc598 [ONNX] remove outdated ImplicitCastType QA in onnx.rst (#81268)
ed1da2a9df [ONNX] Quantization support for quantized::cat (#79826)
773d80747c [ONNX] Clean up unit tests, rename files and improve import style (#81141)
06710ec1b9 [ONNX] Reland: Add quantization support to _avg_pool opset 9 and clean up (#81267)
b98b9eaae5 Revert "[ONNX] Add quantization support to _avg_pool opset 9 and clean up (#79793)"
356341a3ec [ONNX] Add quantization support to _avg_pool opset 9 and clean up (#79793)
80bf2ea3d9 [CI] Install vision without pep517 (#81074)
712b3e76ef [onnx] Add argsort support (#80234)
f608db5ff0 [ONNX] Moved uninitialized_optional test cases into runtime test and parameterized left test cases (#80145)
b00e4e8017 [ONNX] Convert aten numpy_T to ONNX transpose (#79269)
667fdd72c8 Revert "Revert [ONNX] Enable data propagation from ONNX to simplify current shape inference" (#80730)
7846953316 [ONNX] Constant folding to not add folded constant as initializer (#79552)
127ac09455 Revert "[ONNX] Enable data propagation from ONNX to simplify current shape inference (#75307)"
bee2d485d9 [ONNX] Enable data propagation from ONNX to simplify current shape inference (#75307)
d1d2687d34 [ONNX] Fix potentially unbound variables (#79789)
3dec9fd09f [ONNX] Fix `hardshrink` and `softshrink` output's shape (#79695)
7a8d6c9b1d [ONNX] Fix onnx logical functions' dtype (#79339)
33761c80d2 [ONNX] Move no runtime tests out of runtime test cases (#80078)
86682caf93 [ONNX Export] Use `half_pixel` instead of `pytorch_half_pixel`. (#80003)
da33c93169 [ONNX] Clean up `onnx_supported_ops` (#79424)
438142a599 [ONNX] Update onnx submodule to 1.12 (#79585)
e1caf3465b [ONNX] Refactor `TestONNXRuntime` with parameterized class
efbee797b8 [ONNX] Fix `prelu` output's shape (#79846)
4b52babcd9 [ONNX] Fix `any` and `all` outputs' shape (#79371)
c47f78d25f [ONNX] Fix linalg norm output's shapes and dtypes (#79506)
e3c9cb675a Explicitly convert to double for comparison (#79964)
aa911efdeb [ONNX] Fix `_graph_op` type annotation (#79882)
e675c33c51 [onnx] have all classes that call setUp inherit from common_utils.setUp (#79543)
73f6601cfc [ONNX] Refactor heavy memory usage tests
4b817f5816 (tag: ciflow/periodic/78231) [ONNX] Improve docstrings and typing annotations (#78231)
c8b9b6266b [ONNX] Fix arg type in `_set_training_mode` (#78583)
d2fbfe7fce [ONNX] subscribe onnx to our custom test infra (#79546)
28d1216bd5 Skip Flaky ONNX Test (#79556)
44764f131b [ONNX] Move tests in test_onnx_export.py to test_pytorch_onnx_no_runtime.py (#78310)
d3ef5c3fa3 [ONNX] Clean up `__init__` in torch.onnx (#78446)
1a845579b6 [ONNX] Fix inconsistent rand dtype (#79193)
134459161d [ONNX] Improve shape inference supporting inferred values and unspecified optionals (#789
99)
d26c575ff5 [ONNX] Enable script tests based on recent exporter changes (#77254)
3f14ca8f02 [ONNX] Compare types for DeviceObjType instead of strings (#78114)
4750f745bf [ONNX] Disable parallel run for custom op related tests in CI (#78944)
da805953fb Subscribe onnx _BaseTestCase to TestCase setUp() (#78933)
530dcc2b94 [ONNX] Tool to verify exported model discrepancy between sets of inputs
491c2ed281 (upstream/gh/z-a-f/154/base) [ONNX] Use TORCH_WARN for warnings (#78441)
7adf929a99 Skip flaky ONNX test: test_custom_op_fallthrough (#78936)
def778527e [ONNX] Quantization support for five ops (#78103)
0f05e39870 [ONNX] Fix shape inconsistency when exporting scalar log2 (#78701)
fb7a761ffd [ONNX] reduce log spam when exporting dropout in training mode (#78309)
c0814bff87 [ONNX] Variable length argument support for quantized_args (#78775)
cfc968956c [ONNX] Update CI test script to run parallel by default (#78200)
b27f0fea2c [ONNX] Fix case in type annotation in docs (#78388)
6883b0ce9f [ONNX][WIP] Refactor verification.py
299fbbccec [ONNX] Fix `check_training_mode` in symbolic_helper (#78376)
483bb4f0cb [ONNX] Extract export verification as standalone api from unittest
5e03dfd36d [ONNX] Wrap test decorators with functools.wraps (#78254)
161e931156 [ONNX] Modernize python syntax (#77935)
652ecc9ad9 [ONNX] Fix typo when comparing DeviceObjType (#78085)
(END)
```

### PR DESCRIPTION
**Pull request by BowenBao**
Fixes for quantization export to ONNX and a bug that causes incorrectly inferred shapes during export.

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

**Comment by BowenBao** [comment]
@malfet this should be ready once CI is green.

File: test/onnx/test_pytorch_onnx_no_runtime.py [hunk 1]
```diff
@@ -109,6 +110,73 @@ def forward(self, y: Optional[Tensor]) -> Optional[Tensor]:
 
 instantiate_parametrized_tests(TestOptionalOutput)
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_no_runtime.py:111]: We may need to move this to the bottom

**Review by justinchuby — APPROVED** [review]

**Comment by BowenBao** [comment]
@malfet This 1.12.1 pull request for ONNX is ready. Please help us cherry pick.

**Review by malfet — CHANGES_REQUESTED** [review]
What are the corresponding trunk PRs? (Nothing can be cherry-picked into release branch unless it were landed in trunk first)
And what is the justification for those fixes? Is this a regression from 1.11? Or a bug in a newly introduced feature?

**Comment by BowenBao** [comment]
> What are the corresponding trunk PRs? (Nothing can be cherry-picked into release branch unless it were landed in trunk first) And what is the justification for those fixes? Is this a regression from 1.11? Or a bug in a newly introduced feature?

What are trunk PRs? We are following cherry pick steps like #81976. Please let us know if we missed any steps. 
These are important fixes that unblock customer models. Justifications are as follows. 
```
Regression from 1.12:
0b13260 [ONNX] Test and address autograd.func (FusedLayerNorm) shape inference (#81931)

Fixing quantization export introduced in 1.12:
11f7463 Add more functions to __init__.pyi.in for torch._C for Node and Value (#79654)
ed1da2a [ONNX] Quantization support for quantized::cat (#79826)
356341a [ONNX] Add quantization support to _avg_pool opset 9 and clean up (#79793)
def7785 [ONNX] Quantization support for five ops (#78103)
c0814bf [ONNX] Variable length argument support for quantized_args (#78775)
```

**Review by atalman — APPROVED** [review]
Approving as per discussion

**Pull request merged**